src/sknnr_spatial/image.py

```diff
@@ -0,0 +1,346 @@
+        n_outputs = len(output_dims) if output_dims is not None else 1
+
+        return _ImageChunk(
+            # Copy to avoid mutating the original image
+            self.image.copy(),
+            nodata_vals=self.nodata_vals,
+            nan_fill=nan_fill,
+        ).apply(
+            func,
+            returns_tuple=n_outputs > 1,
+            mask_nodata=mask_nodata,
+            **ufunc_kwargs,
+        )
+
+
+class DataArrayImage(Image):
+    band_dim = 0
+
+    def __init__(self, image: xr.DataArray, nodata_vals: NoDataType = None):
+        super().__init__(image, nodata_vals=nodata_vals)
+        self.band_dim_name = image.dims[self.band_dim]
+
+    @property
+    def band_names(self) -> NDArray:
+        return self.image[self.band_dim_name].values
+
+    def _validate_nodata_vals(self, nodata_vals: NoDataType) -> NDArray | None:
+        """
+        Get an array of NoData values in the shape (bands,) based on user input and
+        DataArray metadata.
+        """
+        # Defer to user-provided NoData values over stored attributes
+        if nodata_vals is not None:
+            return super()._validate_nodata_vals(nodata_vals)
+
+        # If present, broadcast the _FillValue attribute to all bands
+        fill_val = self.image.attrs.get("_FillValue")
+        if fill_val is not None:
+            return np.full((self.n_bands,), fill_val)
+
+        return None
+
+    def _postprocess(
+        self,
+        result: xr.DataArray,
+        output_coords: dict[str, list[str | int]],
+    ) -> xr.DataArray:
+        """Process the output of an applied ufunc"""
+        if output_coords is not None:
+            result = result.assign_coords(output_coords)
+        var_dim = list(output_coords.keys())[0]
+
+        # apply_gufunc swaps dimension order, so we need to restore it back to
+        # (band, y, x).
+        return result.transpose(var_dim, ...)
+
+    def apply_ufunc_across_bands(
+        self,
+        func: Callable[Concatenate[NDArray, P], NDArray],
+        *,
+        output_dims: list[list[str]] | None = None,
+        output_dtypes: list[np.dtype] | None = None,
+        output_sizes: dict[str, int] | None = None,
+        output_coords: dict[str, list[str | int]] | None = None,
+        nan_fill: float = 0.0,
+        mask_nodata: bool = True,
+        **ufunc_kwargs,
+    ) -> xr.DataArray | tuple[xr.DataArray]:
+        """
+        Apply a universal function to all bands of the image.
+
+        If the image is backed by a Dask array, the computation will be parallelized
+        across spatial chunks.
+        """
+        image = self.image
+
+        output_dims = output_dims or [["variable"]]
+        n_outputs = len(output_dims)
+        # Fall back to float output if unknown
+        output_dtypes = output_dtypes or [np.float32] * n_outputs
+        # Default to one output band per input band
+        output_sizes = output_sizes or {"variable": len(self.band_names)}
```

Comment:
I went back and forth on the most sensible default for `output_sizes`, which determines the number of outputs in the band dimension. Originally I thought it should be 1, e.g. a single output estimator that would take an image `(256, 256, 8)` and return an image `(256, 256, 1)`, but I changed it last minute to default to the number of input bands, e.g. a basic function like `lambda x: x * 2` that would take an image `(256, 256, 8)` and return an image `(256, 256, 8)`.

I guess a third option would be to not provide *any* default, since that would avoid any possibility of confusion. For reference, if you pass an value that doesn't match what your function actually returns, you won't get an error until you try to compute.

Have any strong feelings on the most reasonable default value here, or whether we should provide a default at all?

Reply:
> (i.e. a "map" rather than a "reduce")

Great way of putting it!

> So maybe I'm leaning toward not providing a default?

I think that's probably reasonable given that there's no "typical" way to use this function. Borrowing your analogy, we could add methods like `map_across_bands` and `reduce_across_bands` with corresponding defaults, but that seems a bit overkill to just avoid entering one parameter, and we would still need `apply_gufunc_across_bands` to handle our `sklearn` methods that generate a different number of outputs than inputs.

I guess the same logic probably applies to `output_dims` and `output_dtypes`, and those defaults should be removed? This *is* a relatively "low-level" piece of `sknnr-spatial` that is mostly for internal use, so making it unsurprising is probably more important than making it easy to use.

> Or does it make sense if the default is None, that we could determine that value with a small compute on a subset of the data?

This is definitely another option, and I know it's at least used by `dask.array.apply_gufunc` when `output_dtypes` isn't given (via [`apply_infer_dtype`](https://github.com/dask/dask/blob/c9f3e39afab6761bc57cfb5c9ed031791212d442/dask/array/core.py#L413)). For some reason, they don't seem to do any inference for the other parameters, which makes me a little hesitant for us to do it. I'd suggest if we go the route of inferring sizes/dimensions/dtypes, that we save that for another PR to avoid making this more unwieldy than I already have!

Reply:
> Borrowing your analogy, we could add methods like `map_across_bands` and `reduce_across_bands` with corresponding defaults, but that seems a bit overkill to just avoid entering one parameter, and we would still need `apply_gufunc_across_bands` to handle our `sklearn` methods that generate a different number of outputs than inputs.

I agree it seems like overkill, especially if it doesn't handle the latter case.  I think `apply` is the correct term here - in thinking to pandas `DataFrame.apply`, it can return a number of different shapes based on the function.  Perhaps that isn't a great analogy, though. 

> I guess the same logic probably applies to `output_dims` and `output_dtypes`, and those defaults should be removed? This _is_ a relatively "low-level" piece of `sknnr-spatial` that is mostly for internal use, so making it unsurprising is probably more important than making it easy to use.

Agreed.  Unsurprising (with examples) seems like the best path forward.  

I might quibble with your thought that this is mostly for internal use, though.  As you've shown with the weighted kNN attribute calculation, this method seems like a pretty key component for doing all kinds of cool stuff with applying functions spatially.  Perhaps you'd point users to `xarray.ufunc_apply` or `dask.gufunc_apply` instead, but having it work seamlessly across different spatial data structures in this package seems big to me!  Again, there might be elements of this that I still don't understand, though.

> I'd suggest if we go the route of inferring sizes/dimensions/dtypes, that we save that for another PR to avoid making this more unwieldy than I already have!

Totally understandable and I'm all for that idea!  This is complicated stuff (to me), so no surprise that it feels unwieldly even though you've done a great job explaining it.

Reply:
I went ahead and removed the defaults for `output_dims`, `output_sizes`, and `output_dtypes` (except for `NDArrayImage.apply_ufunc_across_bands` where we still default to `None` for the unused parameters).

The possibility of inferring parameters is now tracked in #30.